{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5c584d1a-88dd-4107-a2c2-e53edf83633d",
   "metadata": {},
   "source": [
    "## Project Pt 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de8a5100-dfcf-40a1-9afd-49032e5529ef",
   "metadata": {},
   "source": [
    "### 1. Introduction\n",
    "This paper completed a meta-analysis examining the differences in microbiomes of corals with and without stony coral tissue loss disease (sctld). In an effort to reduce biases from varying sampling and bioinformatic methods, they compiled datasets from 16 different studies to identify microbial taxa that may be associated with the disease since the cause of sctld is still unknown. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c1ea9ca-8029-4a6e-9525-6bd13659455d",
   "metadata": {},
   "source": [
    "### 2. Project Pt 2\n",
    "\n",
    "#### 2.1 Contents of Project pt 2\n",
    "Section 1: Data Set 1 from Project Pt1: https://www.nature.com/articles/s43705-023-00220-0\n",
    "- Data sets acquired from 16 studies. **2425** total coral samples (observations). samples collected either from the field or from lab-based experiments and sequenced  \n",
    "- Features (>1000) - taxonomic units (ASVs) that describe how many counts of those ASVs were in each coral sample (the observations)\n",
    "- Metadata about observations - sample type (tissue, mucus, mucus/tissue combo), disease state, geographic location, coral species, aquaria vs natural environment. They also had other data sets of water and sediment samples with ASVs as features. \n",
    "\n",
    "Research Qs:\n",
    "- the cause of stony coral tissue loss disease is unknown but is hypothesized to be from a microbial pathogen - what bacteria are associated with the disease? How do these associated taxa vary across environments and coral species?\n",
    "    - they theorized that a lack of consensus across studies investigating bacteria associated with sctld is from variability in both sample type (location, coral species, field vs aquaria) and bionformatic analysis. They collected data from multiple studies to analysis them using a single bioinformatic pipeline to limit this variability and investigate similar questions by a meta-analysis using the compiled data.\n",
    "- you can also ask how coral microbiomes differ when raised in aquaria vs found in their natural environments. you can also examine how microbiomes of the same species differ across locations and disease states. **comment correction: The data set is useful because it includes microbiome samples for corals raised in aquaria vs those sampled from the field. Since the corals sampled have varying disease states, you can just reframe the question to be about coral microbiomes in different environments instead of focusing on disease state. There is also plenty of samples from the same species which is why this data would be useful in examning microbiomes across disease state and/or location within the same host species.** \n",
    "- they also examined the microbial communities in the water and sediment surrounding some of the coral samples to try to identify potential pathogens that are transmissable \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a16dd84-cf36-4a9a-8ef5-1e3061ec2134",
   "metadata": {
    "tags": []
   },
   "source": [
    "Figure 1: \n",
    "![image.png](attachment:aa9d2e13-d310-4a79-b5d4-d43e7879f975.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6bffe89-146e-4ede-bb01-19e117697fd4",
   "metadata": {},
   "source": [
    "Section 2 - **Figure 1** Answers:\n",
    "- x-axis: Log fold change & relative abundance\n",
    "- y-axis: taxa (genus, order)\n",
    "- A) the data points show the log fold changes in taxonomic groups when comparing disease lesion (diseased tissue sample) to apparently healthy. So, anything showing positive log-fold change means that taxa was more abundant in Disease Lesion, and anything negative was less abundant in Disease Lesion\n",
    "- B) the relative abundances of the taxa groups within the sample types. **comment correction: the relative abundance is calculating the fraction of sample type compared to the rest. the number of samples within a sample type (seawater, mucus, slury, etc) divided by the total number of samples - and the total number of samples is set to 1**\n",
    "- this figure illustrates how bacterial taxa significantly change between healthy and diseased corals, supporting their hypothesis that certain taxa are associated with the disease. B shows how some of these abundances vary across environment or parts of the coral (mucus vs tissue) to support their hypothesis that microbiome changes and associated bacteria may depend on individual hosts. It also shows how disease-associated microbes may be present in the water and sediment, providing evidence for their hypothesis that this disease is transmissable.\n",
    "- Conclusions: Certain bacterial groups are associated with the disease. Multiple genera in Rhodobacterales and Peptostreptococcales increased significantly in diseased samples indicating it may be pathogenic or just associated with the disease state. Some bacteria associated with the disease may be present in the surrounding seawater and sediment, meaning it may be transmissable. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7de98a3c-00e0-4c28-8baa-5c8a054dffa1",
   "metadata": {},
   "source": [
    "Figure 2: ![image.png](attachment:f53cb107-74cb-4be2-ad39-1a5741094d38.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cdf4034-efc0-4f60-843f-792602fd8387",
   "metadata": {},
   "source": [
    "Section 2 - **Figure 2** Answers:\n",
    "- x-axis: Log fold change & relative abundance\n",
    "- y-axis: taxa (genus, order)\n",
    "- A) the data points show the log fold changes in taxonomic groups when comparing disease unaffected (unaffected tissue sample from a diseased coral colony) to apparently healthy. So, anything showing positive log-fold change means that taxa was more abundant in Disease Unaffected, and anything negative was less abundant in Disease Unaffected\n",
    "- B) the relative abundances of the taxa groups within the sample types\n",
    "\n",
    "- this figure also illustrates how bacterial taxa significantly change between healthy and diseased corals, supporting their hypothesis that certain taxa are associated with the disease. This is helpful to provide in addition to Figure 1 because it shows that the microbiome of the entire coral colony changes after being introduced to the disease, even if the disease symptoms haven't spread through the whole coral colony. B shows how some of these abundances vary across environment or parts of the coral (mucus vs tissue) to support their hypothesis that microbiome changes and associated bacteria may depend on individual hosts. It also shows how disease-associated microbes may be present in the water and sediment, providing evidence for their hypothesis that this disease is transmissable.\n",
    "\n",
    "- Conclusions: Both figures together show which bacteria may be associated with healthy corals or the disease. Multiple genera in Rhodobacterales decreased significantly in DU group but Flavobacteriales increased. \n",
    "\n",
    "Another figure that would be helpful would be the same type of figure showing the significant changes in taxa between Disease Unaffected (DU) and Disease Lesion (DL). This would allow you to easily compare all the taxa between all the groups - especially between the disease lesion and disease unaffected. this direct comparison is currently not shown in a figure in the paper. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c06ceab4-61e8-47e4-bb88-b6a793f91d79",
   "metadata": {},
   "source": [
    "Section 3: \n",
    "Data table 1: taxonomy.tsv \n",
    "- tab deliminated data table - opened in excel \n",
    "- 147455 rows & 3 columns. The columns are the feature IDs (which is just the sequence sample identifier), Taxonomy (the identified microbial taxa from the sequence: kingdom - species), and Concensus (percentage that the sequence matched the identified taxa group in the taxonomy database it was matched against). The feature ID's are the observations and the taxonomy and consensus columns are the features. \n",
    "- This data is important because it matches each sequence to a microbial taxa. This enables comparing the taxa between samples which is important to answer the paper's questions. \n",
    "\n",
    "Data table 2: Feature-table.biom\n",
    "- biom table format - downloaded QIIME2 (bioinformatic software package) and accessed through command line: converted biom format to txt file and opened as tab deliminated table in excel\n",
    "- 119456 rows, 2426 columns. rows are feature IDs (match the ones in data table 1), columns are samples (observations). File contains number of feature IDs (sequences of microbial taxa) in each coral sample\n",
    "- this is essential to the paper because knowing the amount of each taxa in the samples allows comparisons between sample groups, so you can identify bacteria associated with diseased coral samples by seeing which taxa are more abundant in diseased samples vs healthy samples.\n",
    "\n",
    "Data table 3: SCTLD_meta_analysis_metadata.csv\n",
    "- comma deliminated table - opened in excel \n",
    "- 2444 rows, 68 columns. rows are sample IDs the columns contain metadata about the samples including the study the data came from, the coral species, the disease state, location of coral. \n",
    "- The file is important to the analysis because it allows you to match the sample ID to the other data tables and contains important information about the samples. Using this in combination with the other data tables, they were able to sort samples into disease state or environment or sample type to compare microbial abundances to answer their research question."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c64b4c04-9e53-47e2-8dd5-aa71cf7cebab",
   "metadata": {},
   "source": [
    "#### 2.2 Project Pt 2\n",
    "- I added why the data set on hand would be useful for other potential research questions and added more information on what relative abundance (one of the axes) is in figure 1. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40e1686a-286d-4020-8e36-6366f2d35da8",
   "metadata": {},
   "source": [
    "### 3. Data Cleaning and Exploration\n",
    "\n",
    "#### 3.1 Load and clean data\n",
    "Link to paper & data:\n",
    "https://www.nature.com/articles/s43705-023-00220-0#data-availability\n",
    "\n",
    "##### Steps to obtain data:\n",
    "1. Click link to github with data \n",
    "https://github.com/srosales712/SCTLD_microbiome_meta_analysis\n",
    "\n",
    "2. Download SCTLD_meta_analysis_metadata.csv and save into ds4bio directory\n",
    "\n",
    "3. Go to Qiime_output folder and download feature-table.biom & taxonomy.tsv and save into ds4bio directory\n",
    "\n",
    "4. Complete code below to convert the \"biom\" format into a tsv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59420f34-e172-4985-9055-4a1f99718432",
   "metadata": {},
   "outputs": [],
   "source": [
    "#install biom-format via command line in terminal \n",
    "pip install biom-format\n",
    "#convert the biom table to a tsv file \n",
    "biom convert -i feature-table.biom -o feature-table.tsv --to-tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac016832-c5bd-4961-bc98-2ed4149489ef",
   "metadata": {},
   "source": [
    "5. Open the feature-table.tsv file in excel (select \"tab-deliminated\" when prompted)\n",
    "\n",
    "6. Open the taxonomy.tsv file in excel the same way \n",
    "\n",
    "7. Open the SCTLD_meta_analysis_metadata.csv in excel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d60f8947-518c-4b2a-84f0-786a27f05323",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Steps to load data in as dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9a721a79-76ce-4bce-a75b-05dbc0e87683",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Run</th>\n",
       "      <th>SRA_accession</th>\n",
       "      <th>Lab</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>Alias</th>\n",
       "      <th>Study</th>\n",
       "      <th>Collection_ID</th>\n",
       "      <th>Species_counts</th>\n",
       "      <th>...</th>\n",
       "      <th>primer_names</th>\n",
       "      <th>forwardPrimer</th>\n",
       "      <th>ReversePrimer</th>\n",
       "      <th>linkerPrimerSequence</th>\n",
       "      <th>platform</th>\n",
       "      <th>platformChemistry</th>\n",
       "      <th>sequencing_facility</th>\n",
       "      <th>barcodes</th>\n",
       "      <th>basepairs</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66NAT</th>\n",
       "      <td>66NAT</td>\n",
       "      <td>SAMN14322012</td>\n",
       "      <td>12</td>\n",
       "      <td>SRS6274849</td>\n",
       "      <td>P_Thome</td>\n",
       "      <td>Pseudodiploria_strigosa</td>\n",
       "      <td>Patriciathome_Pstrigosa</td>\n",
       "      <td>P_Thome</td>\n",
       "      <td>66NAT</td>\n",
       "      <td>High</td>\n",
       "      <td>...</td>\n",
       "      <td>341F_805R</td>\n",
       "      <td>CCTACGGGNGGCWGCAG</td>\n",
       "      <td>GACTACHVGGGTATCTAATCC</td>\n",
       "      <td>TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>CINVESTAV_Merida</td>\n",
       "      <td>NexteraXT</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68NAT</th>\n",
       "      <td>68NAT</td>\n",
       "      <td>SAMN14322013</td>\n",
       "      <td>12</td>\n",
       "      <td>SRS6274852</td>\n",
       "      <td>P_Thome</td>\n",
       "      <td>Pseudodiploria_strigosa</td>\n",
       "      <td>Patriciathome_Pstrigosa</td>\n",
       "      <td>P_Thome</td>\n",
       "      <td>68NAT</td>\n",
       "      <td>High</td>\n",
       "      <td>...</td>\n",
       "      <td>341F_805R</td>\n",
       "      <td>CCTACGGGNGGCWGCAG</td>\n",
       "      <td>GACTACHVGGGTATCTAATCC</td>\n",
       "      <td>TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>CINVESTAV_Merida</td>\n",
       "      <td>NexteraXT</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69NAT</th>\n",
       "      <td>69NAT</td>\n",
       "      <td>SAMN14322014</td>\n",
       "      <td>12</td>\n",
       "      <td>SRS6274848</td>\n",
       "      <td>P_Thome</td>\n",
       "      <td>Pseudodiploria_strigosa</td>\n",
       "      <td>Patriciathome_Pstrigosa</td>\n",
       "      <td>P_Thome</td>\n",
       "      <td>69NAT</td>\n",
       "      <td>High</td>\n",
       "      <td>...</td>\n",
       "      <td>341F_805R</td>\n",
       "      <td>CCTACGGGNGGCWGCAG</td>\n",
       "      <td>GACTACHVGGGTATCTAATCC</td>\n",
       "      <td>TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>CINVESTAV_Merida</td>\n",
       "      <td>NexteraXT</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 69 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          SampleID     BioSample Run SRA_accession      Lab  \\\n",
       "sample-id                                                     \n",
       "66NAT        66NAT  SAMN14322012  12    SRS6274849  P_Thome   \n",
       "68NAT        68NAT  SAMN14322013  12    SRS6274852  P_Thome   \n",
       "69NAT        69NAT  SAMN14322014  12    SRS6274848  P_Thome   \n",
       "\n",
       "                        Experiment                    Alias    Study  \\\n",
       "sample-id                                                              \n",
       "66NAT      Pseudodiploria_strigosa  Patriciathome_Pstrigosa  P_Thome   \n",
       "68NAT      Pseudodiploria_strigosa  Patriciathome_Pstrigosa  P_Thome   \n",
       "69NAT      Pseudodiploria_strigosa  Patriciathome_Pstrigosa  P_Thome   \n",
       "\n",
       "          Collection_ID Species_counts  ... primer_names      forwardPrimer  \\\n",
       "sample-id                               ...                                   \n",
       "66NAT             66NAT           High  ...    341F_805R  CCTACGGGNGGCWGCAG   \n",
       "68NAT             68NAT           High  ...    341F_805R  CCTACGGGNGGCWGCAG   \n",
       "69NAT             69NAT           High  ...    341F_805R  CCTACGGGNGGCWGCAG   \n",
       "\n",
       "                   ReversePrimer  \\\n",
       "sample-id                          \n",
       "66NAT      GACTACHVGGGTATCTAATCC   \n",
       "68NAT      GACTACHVGGGTATCTAATCC   \n",
       "69NAT      GACTACHVGGGTATCTAATCC   \n",
       "\n",
       "                                        linkerPrimerSequence  platform  \\\n",
       "sample-id                                                                \n",
       "66NAT      TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...  Illumina   \n",
       "68NAT      TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...  Illumina   \n",
       "69NAT      TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...  Illumina   \n",
       "\n",
       "          platformChemistry sequencing_facility   barcodes basepairs notes  \n",
       "sample-id                                                                   \n",
       "66NAT                 MiSeq    CINVESTAV_Merida  NexteraXT     250.0   NaN  \n",
       "68NAT                 MiSeq    CINVESTAV_Merida  NexteraXT     250.0   NaN  \n",
       "69NAT                 MiSeq    CINVESTAV_Merida  NexteraXT     250.0   NaN  \n",
       "\n",
       "[3 rows x 69 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#import packages \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "#load in the metadata from csv as a dataframe. index_col=0 to make row names the first column\n",
    "metadata = pd.read_csv(\"SCTLD_meta_analysis_metadata.csv\", index_col = 0)\n",
    "metadata.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "44e3af92-65ca-4e00-951f-0515e66a7845",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Run</th>\n",
       "      <th>SRA_accession</th>\n",
       "      <th>Lab</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>Alias</th>\n",
       "      <th>Study</th>\n",
       "      <th>Collection_ID</th>\n",
       "      <th>Species_counts</th>\n",
       "      <th>collection_time</th>\n",
       "      <th>...</th>\n",
       "      <th>primer_names</th>\n",
       "      <th>forwardPrimer</th>\n",
       "      <th>ReversePrimer</th>\n",
       "      <th>linkerPrimerSequence</th>\n",
       "      <th>platform</th>\n",
       "      <th>platformChemistry</th>\n",
       "      <th>sequencing_facility</th>\n",
       "      <th>barcodes</th>\n",
       "      <th>basepairs</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample-id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66NAT</th>\n",
       "      <td>SAMN14322012</td>\n",
       "      <td>12</td>\n",
       "      <td>SRS6274849</td>\n",
       "      <td>P_Thome</td>\n",
       "      <td>Pseudodiploria_strigosa</td>\n",
       "      <td>Patriciathome_Pstrigosa</td>\n",
       "      <td>P_Thome</td>\n",
       "      <td>66NAT</td>\n",
       "      <td>High</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>341F_805R</td>\n",
       "      <td>CCTACGGGNGGCWGCAG</td>\n",
       "      <td>GACTACHVGGGTATCTAATCC</td>\n",
       "      <td>TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>CINVESTAV_Merida</td>\n",
       "      <td>NexteraXT</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68NAT</th>\n",
       "      <td>SAMN14322013</td>\n",
       "      <td>12</td>\n",
       "      <td>SRS6274852</td>\n",
       "      <td>P_Thome</td>\n",
       "      <td>Pseudodiploria_strigosa</td>\n",
       "      <td>Patriciathome_Pstrigosa</td>\n",
       "      <td>P_Thome</td>\n",
       "      <td>68NAT</td>\n",
       "      <td>High</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>341F_805R</td>\n",
       "      <td>CCTACGGGNGGCWGCAG</td>\n",
       "      <td>GACTACHVGGGTATCTAATCC</td>\n",
       "      <td>TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>CINVESTAV_Merida</td>\n",
       "      <td>NexteraXT</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69NAT</th>\n",
       "      <td>SAMN14322014</td>\n",
       "      <td>12</td>\n",
       "      <td>SRS6274848</td>\n",
       "      <td>P_Thome</td>\n",
       "      <td>Pseudodiploria_strigosa</td>\n",
       "      <td>Patriciathome_Pstrigosa</td>\n",
       "      <td>P_Thome</td>\n",
       "      <td>69NAT</td>\n",
       "      <td>High</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>341F_805R</td>\n",
       "      <td>CCTACGGGNGGCWGCAG</td>\n",
       "      <td>GACTACHVGGGTATCTAATCC</td>\n",
       "      <td>TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>CINVESTAV_Merida</td>\n",
       "      <td>NexteraXT</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              BioSample Run SRA_accession      Lab               Experiment  \\\n",
       "sample-id                                                                     \n",
       "66NAT      SAMN14322012  12    SRS6274849  P_Thome  Pseudodiploria_strigosa   \n",
       "68NAT      SAMN14322013  12    SRS6274852  P_Thome  Pseudodiploria_strigosa   \n",
       "69NAT      SAMN14322014  12    SRS6274848  P_Thome  Pseudodiploria_strigosa   \n",
       "\n",
       "                             Alias    Study Collection_ID Species_counts  \\\n",
       "sample-id                                                                  \n",
       "66NAT      Patriciathome_Pstrigosa  P_Thome         66NAT           High   \n",
       "68NAT      Patriciathome_Pstrigosa  P_Thome         68NAT           High   \n",
       "69NAT      Patriciathome_Pstrigosa  P_Thome         69NAT           High   \n",
       "\n",
       "          collection_time  ... primer_names      forwardPrimer  \\\n",
       "sample-id                  ...                                   \n",
       "66NAT                 NaN  ...    341F_805R  CCTACGGGNGGCWGCAG   \n",
       "68NAT                 NaN  ...    341F_805R  CCTACGGGNGGCWGCAG   \n",
       "69NAT                 NaN  ...    341F_805R  CCTACGGGNGGCWGCAG   \n",
       "\n",
       "                   ReversePrimer  \\\n",
       "sample-id                          \n",
       "66NAT      GACTACHVGGGTATCTAATCC   \n",
       "68NAT      GACTACHVGGGTATCTAATCC   \n",
       "69NAT      GACTACHVGGGTATCTAATCC   \n",
       "\n",
       "                                        linkerPrimerSequence  platform  \\\n",
       "sample-id                                                                \n",
       "66NAT      TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...  Illumina   \n",
       "68NAT      TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...  Illumina   \n",
       "69NAT      TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...  Illumina   \n",
       "\n",
       "          platformChemistry sequencing_facility   barcodes basepairs notes  \n",
       "sample-id                                                                   \n",
       "66NAT                 MiSeq    CINVESTAV_Merida  NexteraXT     250.0   NaN  \n",
       "68NAT                 MiSeq    CINVESTAV_Merida  NexteraXT     250.0   NaN  \n",
       "69NAT                 MiSeq    CINVESTAV_Merida  NexteraXT     250.0   NaN  \n",
       "\n",
       "[3 rows x 68 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Data Cleaning:\n",
    "#from viewing the meta dataframe: index and 1st column are duplicates, so remove first column\n",
    "meta=metadata.drop(\"SampleID\",axis=1)\n",
    "meta.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "294a4105-7ead-4369-910a-e74b69259c2f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxon</th>\n",
       "      <th>Consensus</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>f4f2731649deea1773042cf5d18a0a33</th>\n",
       "      <td>d__Eukaryota; p__Dinoflagellata; c__Dinophycea...</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dc99cb7f143b60159d9626ae5be10656</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82eb6703b24ac1c5e2e7384217cb6747</th>\n",
       "      <td>d__Bacteria; p__Bacteroidota; c__Bacteroidia; ...</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                              Taxon  \\\n",
       "Feature ID                                                                            \n",
       "f4f2731649deea1773042cf5d18a0a33  d__Eukaryota; p__Dinoflagellata; c__Dinophycea...   \n",
       "dc99cb7f143b60159d9626ae5be10656                                         Unassigned   \n",
       "82eb6703b24ac1c5e2e7384217cb6747  d__Bacteria; p__Bacteroidota; c__Bacteroidia; ...   \n",
       "\n",
       "                                  Consensus  \n",
       "Feature ID                                   \n",
       "f4f2731649deea1773042cf5d18a0a33   1.000000  \n",
       "dc99cb7f143b60159d9626ae5be10656   1.000000  \n",
       "82eb6703b24ac1c5e2e7384217cb6747   0.666667  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load in the taxonomy from tsv as a dataframe. index_col=0 to make row names the first column, and it is tab seperated instead of comma\n",
    "taxonomy = pd.read_csv(\"taxonomy.tsv\",sep='\\t', index_col = 0)\n",
    "taxonomy.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "ee874a0f-2a01-4d7e-bdd0-61ee0867c8db",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>f4f2731649deea1773042cf5d18a0a33</th>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Dinoflagellata</td>\n",
       "      <td>Dinophyceae</td>\n",
       "      <td>Gymnodiniphycidae</td>\n",
       "      <td>Suessiaceae</td>\n",
       "      <td>Symbiodinium</td>\n",
       "      <td>Symbiodinium_sp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dc99cb7f143b60159d9626ae5be10656</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82eb6703b24ac1c5e2e7384217cb6747</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Flavobacteriales</td>\n",
       "      <td>Flavobacteriaceae</td>\n",
       "      <td>NS5_marine_group</td>\n",
       "      <td>uncultured_bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d5ca8ad24be354dabb6003c31e6c756e</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteriia</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>Chloroplast</td>\n",
       "      <td>uncultured_bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97174d94588f1829821048d32484b87f</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Cytophagales</td>\n",
       "      <td>Amoebophilaceae</td>\n",
       "      <td>Candidatus_Amoebophilus</td>\n",
       "      <td>uncultured_bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fcec591ea5c151cbdf7cc9fb782b40e4</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>e80ac41e0b4a6019cf3bc006884fea79</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>e817b99fcf0e1156ccb970c5a4505ee0</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fff19592f0161ad7e004ac44adda984e</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ff4bdfe5bdda943ce41f4b028c6a3389</th>\n",
       "      <td>Unassigned</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>147455 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           0                1  \\\n",
       "Feature ID                                                      \n",
       "f4f2731649deea1773042cf5d18a0a33   Eukaryota   Dinoflagellata   \n",
       "dc99cb7f143b60159d9626ae5be10656  Unassigned             None   \n",
       "82eb6703b24ac1c5e2e7384217cb6747    Bacteria     Bacteroidota   \n",
       "d5ca8ad24be354dabb6003c31e6c756e    Bacteria    Cyanobacteria   \n",
       "97174d94588f1829821048d32484b87f    Bacteria     Bacteroidota   \n",
       "...                                      ...              ...   \n",
       "fcec591ea5c151cbdf7cc9fb782b40e4  Unassigned             None   \n",
       "e80ac41e0b4a6019cf3bc006884fea79  Unassigned             None   \n",
       "e817b99fcf0e1156ccb970c5a4505ee0  Unassigned             None   \n",
       "fff19592f0161ad7e004ac44adda984e  Unassigned             None   \n",
       "ff4bdfe5bdda943ce41f4b028c6a3389  Unassigned             None   \n",
       "\n",
       "                                                2                   3  \\\n",
       "Feature ID                                                              \n",
       "f4f2731649deea1773042cf5d18a0a33      Dinophyceae   Gymnodiniphycidae   \n",
       "dc99cb7f143b60159d9626ae5be10656             None                None   \n",
       "82eb6703b24ac1c5e2e7384217cb6747      Bacteroidia    Flavobacteriales   \n",
       "d5ca8ad24be354dabb6003c31e6c756e   Cyanobacteriia         Chloroplast   \n",
       "97174d94588f1829821048d32484b87f      Bacteroidia        Cytophagales   \n",
       "...                                           ...                 ...   \n",
       "fcec591ea5c151cbdf7cc9fb782b40e4             None                None   \n",
       "e80ac41e0b4a6019cf3bc006884fea79             None                None   \n",
       "e817b99fcf0e1156ccb970c5a4505ee0             None                None   \n",
       "fff19592f0161ad7e004ac44adda984e             None                None   \n",
       "ff4bdfe5bdda943ce41f4b028c6a3389             None                None   \n",
       "\n",
       "                                                   4  \\\n",
       "Feature ID                                             \n",
       "f4f2731649deea1773042cf5d18a0a33         Suessiaceae   \n",
       "dc99cb7f143b60159d9626ae5be10656                None   \n",
       "82eb6703b24ac1c5e2e7384217cb6747   Flavobacteriaceae   \n",
       "d5ca8ad24be354dabb6003c31e6c756e         Chloroplast   \n",
       "97174d94588f1829821048d32484b87f     Amoebophilaceae   \n",
       "...                                              ...   \n",
       "fcec591ea5c151cbdf7cc9fb782b40e4                None   \n",
       "e80ac41e0b4a6019cf3bc006884fea79                None   \n",
       "e817b99fcf0e1156ccb970c5a4505ee0                None   \n",
       "fff19592f0161ad7e004ac44adda984e                None   \n",
       "ff4bdfe5bdda943ce41f4b028c6a3389                None   \n",
       "\n",
       "                                                         5  \\\n",
       "Feature ID                                                   \n",
       "f4f2731649deea1773042cf5d18a0a33              Symbiodinium   \n",
       "dc99cb7f143b60159d9626ae5be10656                      None   \n",
       "82eb6703b24ac1c5e2e7384217cb6747          NS5_marine_group   \n",
       "d5ca8ad24be354dabb6003c31e6c756e               Chloroplast   \n",
       "97174d94588f1829821048d32484b87f   Candidatus_Amoebophilus   \n",
       "...                                                    ...   \n",
       "fcec591ea5c151cbdf7cc9fb782b40e4                      None   \n",
       "e80ac41e0b4a6019cf3bc006884fea79                      None   \n",
       "e817b99fcf0e1156ccb970c5a4505ee0                      None   \n",
       "fff19592f0161ad7e004ac44adda984e                      None   \n",
       "ff4bdfe5bdda943ce41f4b028c6a3389                      None   \n",
       "\n",
       "                                                      6  \n",
       "Feature ID                                               \n",
       "f4f2731649deea1773042cf5d18a0a33       Symbiodinium_sp.  \n",
       "dc99cb7f143b60159d9626ae5be10656                   None  \n",
       "82eb6703b24ac1c5e2e7384217cb6747   uncultured_bacterium  \n",
       "d5ca8ad24be354dabb6003c31e6c756e   uncultured_bacterium  \n",
       "97174d94588f1829821048d32484b87f   uncultured_bacterium  \n",
       "...                                                 ...  \n",
       "fcec591ea5c151cbdf7cc9fb782b40e4                   None  \n",
       "e80ac41e0b4a6019cf3bc006884fea79                   None  \n",
       "e817b99fcf0e1156ccb970c5a4505ee0                   None  \n",
       "fff19592f0161ad7e004ac44adda984e                   None  \n",
       "ff4bdfe5bdda943ce41f4b028c6a3389                   None  \n",
       "\n",
       "[147455 rows x 7 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Data Cleaning:\n",
    "#separate the different phylogenetic classifications (in the Taxon column) into different columns by the deliminater ;. expand=True makes it expand the results into different columns  \n",
    "taxonomy_levels=taxonomy['Taxon'].str.split(';', expand=True)\n",
    "\n",
    "#remove the prefixes in the classifications in all taxa columns \n",
    "#https://stackoverflow.com/questions/26837998/pandas-replace-nan-with-blank-empty-string\n",
    "taxonomy_levels2 = taxonomy_levels.replace('[a-z]__', '', regex=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "42d9d5cc-f79e-457e-9da2-bcc74ef79d48",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxon</th>\n",
       "      <th>Consensus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>d__Eukaryota; p__Dinoflagellata; c__Dinophycea...</td>\n",
       "      <td>0.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Taxon  Consensus\n",
       "0  d__Eukaryota; p__Dinoflagellata; c__Dinophycea...        0.8\n",
       "1  d__Bacteria; p__Proteobacteria; c__Alphaproteo...        0.9"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Sample dataframe\n",
    "data = {'Taxon': ['d__Eukaryota; p__Dinoflagellata; c__Dinophyceae; o__Gymnodiniphycidae; f__Suessiaceae; g__Symbiodinium; s__Symbiodinium_sp.',\n",
    "                  'd__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rhodobacterales; f__Rhodobacteraceae; g__Rhodobacter; s__Rhodobacter_sp.'],\n",
    "        'Consensus': [0.8, 0.9]}\n",
    "\n",
    "df = pd.DataFrame(data)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b8005e7a-3939-41ad-9599-45273a507fd5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>d__Eukaryota</td>\n",
       "      <td>p__Dinoflagellata</td>\n",
       "      <td>c__Dinophyceae</td>\n",
       "      <td>o__Gymnodiniphycidae</td>\n",
       "      <td>f__Suessiaceae</td>\n",
       "      <td>g__Symbiodinium</td>\n",
       "      <td>s__Symbiodinium_sp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Alphaproteobacteria</td>\n",
       "      <td>o__Rhodobacterales</td>\n",
       "      <td>f__Rhodobacteraceae</td>\n",
       "      <td>g__Rhodobacter</td>\n",
       "      <td>s__Rhodobacter_sp.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0                  1                       2  \\\n",
       "0  d__Eukaryota  p__Dinoflagellata          c__Dinophyceae   \n",
       "1   d__Bacteria  p__Proteobacteria  c__Alphaproteobacteria   \n",
       "\n",
       "                      3                    4                5  \\\n",
       "0  o__Gymnodiniphycidae       f__Suessiaceae  g__Symbiodinium   \n",
       "1    o__Rhodobacterales  f__Rhodobacteraceae   g__Rhodobacter   \n",
       "\n",
       "                     6  \n",
       "0  s__Symbiodinium_sp.  \n",
       "1   s__Rhodobacter_sp.  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Split the 'Taxon' column by semicolons, expand into separate columns\n",
    "taxonomy_levels = df['Taxon'].str.split('; ', expand=True)\n",
    "taxonomy_levels\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b200185c-f12c-4ffb-aed1-d971668873cc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Consensus          0               1                    2  \\\n",
      "0        0.8  Eukaryota  Dinoflagellata          Dinophyceae   \n",
      "1        0.9   Bacteria  Proteobacteria  Alphaproteobacteria   \n",
      "\n",
      "                   3                 4             5                 6  \n",
      "0  Gymnodiniphycidae       Suessiaceae  Symbiodinium  Symbiodinium_sp.  \n",
      "1    Rhodobacterales  Rhodobacteraceae   Rhodobacter   Rhodobacter_sp.  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xz/j1bn_g_96wjd99l2tvy0cdq40000gn/T/ipykernel_7655/2845537776.py:2: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  taxonomy_levels = taxonomy_levels.apply(lambda x: x.str.replace('[a-z]__', ''))\n"
     ]
    }
   ],
   "source": [
    "# Remove prefixes from each column\n",
    "taxonomy_levels = taxonomy_levels.apply(lambda x: x.str.replace('[a-z]__', ''))\n",
    "\n",
    "# Combine the taxonomy levels with the original dataframe\n",
    "df = pd.concat([df, taxonomy_levels], axis=1)\n",
    "\n",
    "# Drop the original 'Taxon' column if needed\n",
    "df = df.drop('Taxon', axis=1)\n",
    "\n",
    "# Display the updated dataframe\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "042a430f-b34f-41ba-8855-9b95c9c5bb16",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
